kernel/src/main/java/org/kframework/parser/concrete2kore/disambiguation/TypeInferencer.java

```diff
@@ -0,0 +1,784 @@
+                  || tc.production().klabel().get().name().startsWith("#SemanticCastTo")
+                  || tc.production().klabel().get().name().equals("#InnerCast"))) {
+              expectedSort = getSortOfCast(tc);
+              isStrictEquality = tc.production().klabel().get().name().equals("#SyntacticCast")
+                  || tc.production().klabel().get().name().equals("#InnerCast");
+              ids.add(apply(tc.get(j)));
+            } else if (isTopSort && j == 0 && isFunction(tc.get(j), isAnywhere)) {
+              expectedSort = getFunctionSort(tc.get(j));
+              expectedParams = Optional.of(getFunction(tc.get(j)).get());
+              ids.add(apply(tc.get(j)));
+            } else {
+              expectedSort = nt.sort();
+              ids.add(apply(tc.get(j)));
+            }
+            j++;
+          }
+        }
+        isStrictEquality = wasStrict;
+        expectedSort = oldExpectedSort;
+        expectedParams = oldExpectedParams;
+      }
+      String expected = printSort(expectedSort, expectedParams, false).replace("|", "");
+      boolean cached = !cacheById.get(id).add(expected);
+      if (!isIncremental && (!shared || !cached)) {
+        sb.append("(define-fun |constraint").append(id).append("_").append(expected).append("| () Bool (and true ");
+      }
+      if (isIncremental || !shared || !cached) {
+        if (pr instanceof Constant && (pr.production().sort().equals(Sorts.KVariable()) || pr.production().sort().equals(Sorts.KConfigVar()))) {
+          Constant c = (Constant) pr;
+          String name;
+          if (!shared) {
+            nextId++;
+            variablesById.add(new ArrayList<>());
+            cacheById.add(new HashSet<>());
+            pr.setId(Optional.of(id));
+            if (isAnonVar(c)) {
+              name = "Var" + c.value() + locStr(c);
+            } else {
+              name = "Var" + c.value();
+            }
+            if (!variables.contains(name)) {
+              variables.add(name);
+              variableNames.add(c.value());
+            }
+            variablesById.get(id).add(name);
+          } else {
+            name = variablesById.get(id).get(0);
+          }
+          Sort actualSort = null;
+          pushConstraint(name, c);
+        } else if (isRealSort(pr.production().sort())) {
+          pushConstraint(pr.production().sort(), Optional.of(pr));
+        }
+      }
+      if (!isIncremental && (!shared ||  !cached)) {
+        for (String i : ids) {
+          sb.append(i).append(" ");
+        }
+        sb.append("))\n");
+      }
+      return "|constraint" + id + "_" + expected + "|";
+    }
+
+    public boolean isAnonVar(Constant var) {
+      return var.value().equals(ResolveAnonVar.ANON_VAR.name()) || var.value().equals(ResolveAnonVar.FRESH_ANON_VAR.name());
+    }
+
+    private void pushConstraint(Sort actualSort, Optional<ProductionReference> actualParams) {
```

Comment:
If I understand correctly, this pushes a sort constraint, and refers to the sort relations which have already been defined?

Reply:
yes